### PR TITLE
Fixes for annotations bugs.

### PR DESCRIPTION
**Pull request by paulp**
This supersedes and renders irrelevant #1802.

### CODE REVIEW

**Comment by paulp** [comment]
Review by @lrytz if available, else @retronym wins the nomination.

File: src/library/scala/throws.scala [hunk 1]
```diff
@@ -24,5 +24,5 @@ package scala
  * @since   2.1
  */
 class throws[T <: Throwable](cause: String = "") extends scala.annotation.StaticAnnotation {
-  def this(clazz: Class[T]) = this()
+  def this(clazz: Class[T]) = this("")
```
Comment (xeno-by) [inline src/library/scala/throws.scala:27]: Why is this change necessary?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -112,6 +112,10 @@ trait Typers extends Modes with Adaptations with Tags {
       }
     }
 
+    private def mkNamedArg(tree: Tree, name: Name): Tree = {
+      atPos(tree.pos)(new AssignOrNamedArg(Ident(name), tree))
+    }
+
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:118]: Maybe let's move this to `TreeGen`?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 3]
```diff
@@ -3437,38 +3443,35 @@ trait Typers extends Modes with Adaptations with Tags {
...
-              extract(f, args :: outerArgss)
-            case Select(New(tpt), nme.CONSTRUCTOR) =>
-              (fun, outerArgss)
-            case _ =>
-              reportAnnotationError(UnexpectedTreeAnnotation(fun))
-              (setError(fun), outerArgss)
-          }
-        extract(ann, List())
-      }
+      val treeInfo.Applied(fun0, targs, argss) = treeInfo.dissectApplied(ann)
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3446]: This can be written as just `val treeInfo.Applied(fun0, targs, argss) = ann`.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 4]
```diff
@@ -3437,38 +3443,35 @@ trait Typers extends Modes with Adaptations with Tags {
...
+        if (targs.isEmpty && typedFun0.exists(t => isDummyAppliedType(t.tpe)))
+          logResult(s"Retyped $typedFun0 to find type args")(typed(argss.foldLeft(fun0)(Apply(_, _))))
+        else
+          typedFun0
+      )
+      val typedFun @ Select(New(annTpt), _) = treeInfo.dissectApplied(typedFunPart).core
+      val annType = annTpt.tpe
 
-      val res = if (fun.isErroneous) annotationError
+      val res = if (typedFun.isErroneous) annotationError
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3460]: Previously `typed(fun0)` wasn't called if `isErroneous` happened to be `true`. Isn't the current approach going to sometimes cause spurious errors?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 5]
```diff
@@ -3437,38 +3443,35 @@ trait Typers extends Modes with Adaptations with Tags {
...
       // begin typedAnnotation
-      val (fun, argss) = {
-        def extract(fun: Tree, outerArgss: List[List[Tree]]):
-          (Tree, List[List[Tree]]) = fun match {
-            case Apply(f, args) =>
-              extract(f, args :: outerArgss)
-            case Select(New(tpt), nme.CONSTRUCTOR) =>
-              (fun, outerArgss)
-            case _ =>
-              reportAnnotationError(UnexpectedTreeAnnotation(fun))
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3448]: Looks like `UnexpectedTreeAnnotation` is no longer reported for non-classfile annotations?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 6]
```diff
@@ -3437,38 +3443,35 @@ trait Typers extends Modes with Adaptations with Tags {
...
+      val typedFunPart = (
+        // If there are dummy type arguments in typeFun part, it suggests we
+        // must type the actual constructor call, not only the select. The value
+        // arguments are how the type arguments will be inferred.
+        if (targs.isEmpty && typedFun0.exists(t => isDummyAppliedType(t.tpe)))
+          logResult(s"Retyped $typedFun0 to find type args")(typed(argss.foldLeft(fun0)(Apply(_, _))))
+        else
+          typedFun0
+      )
+      val typedFun @ Select(New(annTpt), _) = treeInfo.dissectApplied(typedFunPart).core
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3457]: This is a matter of taste, but probably it'd look more concise as `val treeInfo.Applied(typedFun @ Select(New(annTpt), _), _, _) = typedFunPart`.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 7]
```diff
@@ -3437,38 +3443,35 @@ trait Typers extends Modes with Adaptations with Tags {
...
+            reportAnnotationError(AnnotationTypeMismatchError(annTpt, annType, annType))
           } else if (argss.length > 1) {
             reportAnnotationError(MultipleArgumentListForAnnotationError(ann))
           } else {
-            val args =
-              if (argss.head.length == 1 && !isNamed(argss.head.head))
-                List(new AssignOrNamedArg(Ident(nme.value), argss.head.head))
-              else argss.head
+            val args = argss match {
+              case (arg :: Nil) :: Nil if !isNamed(arg) => mkNamedArg(arg, nme.value) :: Nil
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3472]: What do you think about renaming `isNamed` to `isNamedArg` for consistency?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 8]
```diff
@@ -3524,7 +3527,7 @@ trait Typers extends Modes with Adaptations with Tags {
             // sometimes does. The problem is that "self" ident's within
             // annot.constr will retain the old symbol from the previous typing.
             val func     = Function(funcparm :: Nil, ann.duplicate)
-            val funcType = appliedType(FunctionClass(1), selfsym.info, annClass.tpe_*)
+            val funcType = appliedType(FunctionClass(1), selfsym.info, annType)
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3530]: Using `annType` here is necessary to propagate the info about the inferred type, rght?

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3530]: Also it looks like `annClass` is only used once now. Maybe we can get rid of it?

File: src/reflect/scala/reflect/internal/Types.scala [hunk 9]
```diff
@@ -2915,12 +2915,19 @@ trait Types extends api.Types { self: SymbolTable =>
     else existentialAbstraction(existentialsInType(tp), tp)
   )
 
-  def containsExistential(tpe: Type) =
-    tpe exists typeIsExistentiallyBound
+  def containsDummyTypeArg(tp: Type) = tp exists isDummyTypeArg
+  def isDummyTypeArg(tp: Type) = tp.typeSymbol.isTypeParameter
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Types.scala:2919]: What if it's `T` as in `def foo[T] = weakTypeOf[T]`? Does it still qualify for being a dummy type arg?

File: src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala [hunk 10]
```diff
@@ -423,6 +423,9 @@ trait ContextErrors {
       def UnexpectedTreeAnnotation(tree: Tree) =
         NormalTypeError(tree, "unexpected tree in annotation: "+ tree)
 
+      def AnnotationRequiresTypeArgs(tree: Tree, unapplied: Type) =
+        NormalTypeError(tree, s"$unapplied requires explicit type arguments")
+
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala:428]: Is this code used?

**Comment by JamesIry** [comment]
ping @paulp

File: src/library/scala/throws.scala (hunk 1 above)
Comment (paulp) [inline src/library/scala/throws.scala:27]: It isn't. But I often see how the compiler strains when it comes to default arguments, and sometimes I have pity and try to lessen its burden.

**Comment by paulp** [comment]
Superseded by #1901.
